### PR TITLE
Add clone_message() function

### LINKED ISSUE #138: Duplicating a message with create_message() loses all fields
### Problem description

I have a use case, where I need to remove a sensitive field from a log message, while still maintaining the content of the original message (this is because there is certain information that only privileged users at my company are allowed to see). In a message processing pipeline, I'd like to be able to duplicate a message and all its fields, routing the original copy to one stream, and the filtered copy to another, unprivileged stream. 

However, when using `create_message()`, all fields from the original message are lost, making field removal on the new message impossible.

### Steps to reproduce the problem

1. Enable the Random HTTP Message Generator input
2. Configure two destination streams, one for unfiltered messages, another for filtered.
3. Set up a pipeline processing rule that looks like:

```
rule "filter user ids"
when
  has_field("user_id")
then
  route_to_stream(id: "<stream id 1>");
  let x = create_message();
  remove_field("user_id", x);
  route_to_stream(id: "<stream id 2>", message: x);
end
```

My expectation is that the message created and stored in the `x` variable has all the fields that the original message had, so I can `remove_field()` on the user_id field. But, the new message only has a message, source, and timestamp field. Perhaps another function (i.e. `duplicate_message()`) is desireable.

The ultimate goal here is to enable field level access control -- so another way to do this that doesn't involve duplicating messages in the ES index would be great!

### Environment

* Graylog Version: `2.1.2+50e449a`
* Pipeline Processor plugin version: Unsure of how to determine this
* Elasticsearch Version: `2.4.2`
* MongoDB Version: `2.6.10`
* Operating System: Ubuntu 16.04.1 LTS
* Browser version:

### PR DESCRIPTION
**Pull request by joschi**
Closes #138

### CODE REVIEW

File: plugin/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/CloneMessage.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        // register in context so the processor can inject it later on
+        context.addCreatedMessage(clonedMessage);
+        return clonedMessage;
+    }
+
+    @Override
+    public FunctionDescriptor<Message> descriptor() {
+        return FunctionDescriptor.<Message>builder()
+                .name(NAME)
+                .params(new ParameterDescriptor[0])
```
Comment (kroepke) [inline plugin/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/CloneMessage.java]: Can we add an optional `message` parameter here (analogous to https://github.com/Graylog2/graylog-plugin-pipeline-processor/blob/master/plugin/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RemoveField.java#L34)?

That way we could support cloning other messages than the one from the context, as well:
```
let x = create_message("foo", "source");
let y = clone_message(x);
let z = clone_message();
```

File: plugin/src/test/java/org/graylog/plugins/pipelineprocessor/functions/FunctionsSnippetsTest.java [hunk 2]
```diff
@@ -392,17 +394,47 @@ public void evalErrorSuppressed() {
...
-        final EvaluationContext context = contextForRuleEval(rule, new Message("test", "test", Tools.nowUTC()));
+        final EvaluationContext context = contextForRuleEval(rule, message);
 
         final Message origMessage = context.currentMessage();
         final Message newMessage = Iterables.getOnlyElement(context.createdMessages());
 
         assertThat(origMessage).isNotSameAs(newMessage);
+        assertThat(newMessage.getMessage()).isEqualTo("new");
+        assertThat(newMessage.getSource()).isEqualTo("synthetic");
+        assertThat(newMessage.getStreams()).isEmpty();
```
Comment (kroepke) [inline plugin/src/test/java/org/graylog/plugins/pipelineprocessor/functions/FunctionsSnippetsTest.java]: :+1:

File: plugin/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/CloneMessage.java (hunk 1 above)
Comment (joschi) [inline plugin/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/CloneMessage.java]: ✅ Totally makes sense.

**Comment by kroepke** [comment]
lgtm!

**Pull request merged**